File: 1942-the-number-of-the-smallest-unoccupied-chair/1942-the-number-of-the-smallest-unoccupied-chair.java
class Solution {
    public int smallestChair(int[][] times, int targetFriend) {
        PriorityQueue<int[]> pq = new PriorityQueue<>(
            (a, b) -> a[0]-b[0]
        );

        TreeSet<Integer> freeChair = new TreeSet<>();
        int targetArrivalTime = times[targetFriend][0];

        Arrays.sort(times, (a, b) -> a[0] - b[0]);

        int chair = 0, targetChair = 0;
        for(int[] time : times){
            int arrival = time[0];
            int departure = time[1];

            while(!pq.isEmpty() && pq.peek()[0] <= arrival){
                freeChair.add(pq.poll()[1]);
            }

            int assignedChair = 0;
            if(freeChair.isEmpty()){
                assignedChair = chair++;
            }else{
                assignedChair = freeChair.first();
                freeChair.remove(assignedChair);
            }

            if(arrival == targetArrivalTime) return assignedChair;

            pq.offer(new int[]{departure, assignedChair});
        }

        return -1;
    }
}

File: 1942-the-number-of-the-smallest-unoccupied-chair/1942-the-number-of-the-smallest-unoccupied-chair.cpp
class Solution {
public:
    int smallestChair(vector<vector<int>>& times, int targetFriend) {
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
        set<int> freeChair;
        int targetArrivalTime = times[targetFriend][0];

        sort(times.begin(), times.end());

        int chair = 0, targetChair = 0;
        for(int i = 0; i < times.size(); i++){
            int arrival = times[i][0];
            int departure = times[i][1];

            while(!pq.empty() && pq.top().first <= arrival){
                freeChair.insert(pq.top().second);
                pq.pop();
            }

            int assignedChair = 0;
            if(freeChair.empty()){
                assignedChair = chair++;
            }else{
                assignedChair = *freeChair.begin();
                freeChair.erase(freeChair.begin());
            }

            if(arrival == targetArrivalTime) return assignedChair;

            pq.push({departure, assignedChair});
        }

        return -1;
    }
};